File: website-mongo/src/common/apis/MetroAPI.ts
import ky, { KyResponse } from 'ky';

export class MetroAPI {
  baseURL: string;

  constructor(baseURL: string) {
    this.baseURL = baseURL;
  }

  async metros(): Promise<KyResponse> {
    return ky.get(this.baseURL + '/metros');
  }

  async getMetro(id: number): Promise<KyResponse> {
    return ky.get(this.baseURL + '/metros/' + id);
  }

  async insertMetro(name: string, extendedName: string, metroSizeRank: number, population: number, featuredImage: string, notes: string): Promise<KyResponse> {
    const formData = new FormData();
    formData.set('name', name);
    formData.set('extended_name', extendedName);
    formData.set('metro_size_rank', metroSizeRank.toString());
    formData.set('featured_image', featuredImage);
    formData.set('population', population.toString());
    formData.set('notes', notes);
    return ky.post(this.baseURL + '/metros', { body: formData });
  }

  async editMetro(id: string, name: string, extendedName: string, metroSizeRank: number, population: number, featuredImage: string, notes: string): Promise<KyResponse> {
    const formData = new FormData();
    formData.set('name', name);
    formData.set('extended_name', extendedName);
    formData.set('notes', notes);
    formData.set('metro_size_rank', metroSizeRank.toString());
    formData.set('featured_image', featuredImage);
    formData.set('population', population.toString());
    return ky.put(this.baseURL + '/metros/' + id, { body: formData });
  }

  async deleteMetro(id: number): Promise<KyResponse> {
    return ky.delete(this.baseURL + '/metros/' + id);
  }

}

File: android/app/src/main/java/com/adhiravishankar/citiesorganizer/apis/MetroAPI.java
package com.adhiravishankar.citiesorganizer.apis;

import com.adhiravishankar.citiesorganizer.models.DetailedMetro;
import com.adhiravishankar.citiesorganizer.models.Metro;

import java.util.List;

import retrofit2.Call;
import retrofit2.http.DELETE;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;

public interface MetroAPI {

    @GET("metros")
    Call<List<Metro>> getMetros();

    @FormUrlEncoded
    @POST("metros")
    Call<String> postMetro(@Field("name") String name,
                           @Field("extended_name") String extendedName,
                           @Field("small_name") String smallName,
                           @Field("population") Integer population,
                           @Field("metro_size_rank") Integer metroSizeRank,
                           @Field("notes") String notes,
                           @Field("featured_image") String featuredImage);

    @GET("metros/{metro}")
    Call<DetailedMetro> getMetro(@Path("metro") String metro);

    @FormUrlEncoded
    @PUT("metros/{metro}")
    Call<String> updateMetro(@Path("metro") String metro,
                             @Field("name") String name,
                             @Field("extended_name") String extendedName,
                             @Field("small_name") String smallName,
                             @Field("population") Integer population,
                             @Field("metro_size_rank") Integer metroSizeRank,
                             @Field("notes") String notes,
                             @Field("featured_image") String featuredImage);

    @DELETE("metros/{metro}")
    Call<String> deleteMetro(@Path("metro") String metro);
}
